### PR TITLE
 Create Undo system

### PR DESCRIPTION
**Pull request by ninjrdevelop**
Have recreated the Undo system from the old code. Works for:

- Props
- DynShapes
- Wheel (but it currently doesn't remove the extra GameObject on the parent)
- Weld
- Redo functionality is a TODO.

A couple of other small (mostly related) changes:

- Weld tool secondary (Unweld all) moved to Reload
- Added SteamId, GroupId, and User Name to 'Player' class

### CODE REVIEW

File: Code/Components/UndoSystem/UndoSystem.cs [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+	{
+		var player = Player.FindLocalPlayer();
+		if ( !player.IsValid() )
+			return;
+
+		Undo undo = GetFirstAndRemove( player.SteamId );
+
+		if (undo != null)
+		{
+			if ( undo.UndoCallback != null )
```
Comment (Nebual) [inline Code/Components/UndoSystem/UndoSystem.cs:62]: 🤔 Maybe we should make the only public constructor of `Undo` take a mandatory callback - I feel like thats probably going to be what we always do anyway, and then we can assume its non-null.

File: Code/Components/UndoSystem/UndoSystem.cs [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+			{
+				var undoMessage = undo.UndoCallback();
+				if ( undoMessage != "" )
+				{
+					HintFeed.AddHint( "", undoMessage );
+					CreateUndoParticles( Vector3.Zero );
```
Comment (Nebual) [inline Code/Components/UndoSystem/UndoSystem.cs:68]: Erm, so this never creates particles currently ehh?

I guess for us to have a position, we'd need a separate-from-the-callback reference to the object being undone (or at least object having something about it, like a weld, undone). That appears to be what the original code used `Undo.Prop` for, though perhaps lets generalize it to GameObject?

File: Code/Components/UndoSystem/UndoSystem.cs [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+	}
+
+	public static void Add( Player creator, Func<string> callback, bool avoid = false )
+	{
+		if ( creator == null ) return;
+
+		var undo = new Undo( creator )
+		{
+			UndoCallback = callback,
+			Avoid = avoid
```
Comment (Nebual) [inline Code/Components/UndoSystem/UndoSystem.cs:81]: `Avoid` was a confusingly bad name for this in the old system haha, I had to go back and read it to figure out: this should probably be called `IsHidden` or something _if_ we do the rather-overcomplicated "when undo'ing props, first hide them for a few seconds, so they can be 'redo'd during that window" which.......... seems like functionality that is inevitably going to be a bit glitchy with constraints/etc, even if it looks really impressive.

Perhaps for now, lets just delete `Avoid` from the system, until we make a redo?

File: Libraries/sbox-libevents
Comment (Nebual) [inline Libraries/sbox-libevents:1]: Did you mean to remove this too?

Good idea on adding it as a submodule (I was too sleepy last night to think of that haha), that makes way more sense than committing the code directly like they seem to recommend. Though maybe we'll want to use [this non-library version](https://sbox.game/dev/doc/the-scene-system/events/?q=Event)? it looks a little more powerful

Comment (ninjrdevelop) [reply Libraries/sbox-libevents:1]: I had added it as I used an event to begin with, and then ended up not using it so I stripped it out.

We should probably see how we go with the built in version before outsourcing it 😀

File: Code/Components/UndoSystem/UndoSystem.cs (hunk 2 above)
Comment (ninjrdevelop) [reply Code/Components/UndoSystem/UndoSystem.cs:68]: Yeah I tried to separate out the need for undo to rely on Game Objects, but this is the limitation of that.

I'll readd an optional GO to enable the particle effect.

**Review by Nebual — COMMENTED** [review]
Go ahead and delete Undo.cs and Undoer.cs from CodeOld/ (and maybe if we're not going to bother with the redo system anytime soon, the whole undosystem folder)

File: Code/Components/UndoSystem/UndoSystem.cs (hunk 3 above)
Comment (ninjrdevelop) [reply Code/Components/UndoSystem/UndoSystem.cs]: Removed in 84fbdfb

File: Code/Components/UndoSystem/UndoSystem.cs (hunk 2 above)
Comment (ninjrdevelop) [reply Code/Components/UndoSystem/UndoSystem.cs]: Sorted in 84fbdfb

File: Code/Components/UndoSystem/UndoSystem.cs (hunk 1 above)
Comment (ninjrdevelop) [reply Code/Components/UndoSystem/UndoSystem.cs:62]: Sorted in 84fbdfb

File: Code/Components/UndoSystem/UndoSystem.cs (hunk 2 above)
[the remark this answers is not in the document]
Comment (ninjrdevelop) [reply Code/Components/UndoSystem/UndoSystem.cs]: Love it. dbb2f15

**Comment by ninjrdevelop** [comment]
Good to go @Nebual . Merry Christmas :)

**Review by Nebual — APPROVED** [review]
Excellent, thanks for getting this working!

Merry Christmas to you too :)

**Pull request merged**